code/modules/reagents/chemistry/holder.dm

```diff
@@ -485,44 +558,261 @@
+					remove_reagent(B, (multiplier * cached_required_reagents[B]), safety = 1, ignore_pH = TRUE)
+
+				for(var/P in selected_reaction.results)
+					multiplier = max(multiplier, 1) //this shouldnt happen ...
+					SSblackbox.record_feedback("tally", "chemical_reaction", cached_results[P]*multiplier, P)//log
+					add_reagent(P, cached_results[P]*multiplier, null, chem_temp)
+
+
+				var/list/seen = viewers(4, get_turf(my_atom))//Sound and sight checkers
+				var/iconhtml = icon2html(cached_my_atom, seen)
+				if(cached_my_atom)
+					if(!ismob(cached_my_atom)) // No bubbling mobs
+						if(selected_reaction.mix_sound)
+							playsound(get_turf(cached_my_atom), selected_reaction.mix_sound, 80, 1)
+
+
 						for(var/mob/M in seen)
-							to_chat(M, "<span class='notice'>[iconhtml] \The [my_atom]'s power is consumed in the reaction.</span>")
-							ME2.name = "used slime extract"
-							ME2.desc = "This extract has been used up."
+							to_chat(M, "<span class='notice'>[iconhtml] [selected_reaction.mix_message]</span>")
+
+					if(istype(cached_my_atom, /obj/item/slime_extract))//if there's an extract and it's used up.
+						var/obj/item/slime_extract/ME2 = my_atom
+						ME2.Uses--
+						if(ME2.Uses <= 0) // give the notification that the slime core is dead
+							for(var/mob/M in seen)
+								to_chat(M, "<span class='notice'>[iconhtml] \The [my_atom]'s power is consumed in the reaction.</span>")
+								ME2.name = "used slime extract"
+								ME2.desc = "This extract has been used up."
 
-			selected_reaction.on_reaction(src, multiplier)
-			reaction_occurred = 1
+				selected_reaction.on_reaction(src, multiplier, special_react_result)
+				reaction_occurred = 1
+				continue_reacting = TRUE
 
 	while(reaction_occurred)
 	update_total()
 	return 0
 
+/datum/reagents/process()
+	var/datum/chemical_reaction/fermi/C = fermiReactID
+
+	var/list/cached_required_reagents = C.required_reagents//update reagents list
+	var/list/cached_results = C.results//resultant chemical list
+	var/multiplier = INFINITY
+
+	for(var/B in cached_required_reagents) //
+		multiplier = min(multiplier, round((get_reagent_amount(B) / cached_required_reagents[B]), 0.001))
+	if (multiplier == 0)
+		fermiEnd()
+		return
+	for(var/P in cached_results)
+		targetVol = cached_results[P]*multiplier
+
+	if (fermiIsReacting == FALSE)
+		CRASH("Fermi has refused to stop reacting even though we asked her nicely.")
+
+	if (chem_temp > C.OptimalTempMin && fermiIsReacting)//To prevent pointless reactions
+		if( (pH >= (C.OptimalpHMin - C.ReactpHLim)) && (pH <= (C.OptimalpHMax + C.ReactpHLim)) )
+			if (reactedVol < targetVol)
+				reactedVol = fermiReact(fermiReactID, chem_temp, pH, reactedVol, targetVol, cached_required_reagents, cached_results, multiplier)
+			else//Volume is used up
+				fermiEnd()
+				return
+		else//pH is out of range
+			fermiEnd()
+			return
+	else//Temperature is too low, or reaction has stopped.
+		fermiEnd()
+		return
```

Comment:
needs to be re-written so that it early returns when conditions aren't met instead of meganesting